## Problema de Negócio

### Identificar clientes insatisfeitos com o banco Santander antes que eles encerrem suas contas no banco.

In [1]:
# Fonte dos dados
# https://www.kaggle.com/competitions/santander-customer-satisfaction/data

In [2]:
# Importando Bibliotecas Necessárias
import numpy as np
import pandas as pd
%matplotlib inline

import warnings
warnings.filterwarnings("ignore")

In [3]:
# Versões dos pacotes usados neste jupyter notebook
%reload_ext watermark
%watermark -a "Gilberto Barbosa dos Santos" --iversions

Author: Gilberto Barbosa dos Santos

pandas: 1.4.2
numpy : 1.22.4



In [4]:
# Importando os dados de treinamento
df = pd.read_csv("train.csv", sep = ",", encoding_errors='ignore')

In [5]:
# Mostrar todas as colunas do dataframe
pd.set_option('display.max_columns', None)

In [6]:
# Primeiras linhas do dataset
df.head()

ID  var3  var15  imp_ent_var16_ult1  imp_op_var39_comer_ult1  \
0   1     2     23                 0.0                      0.0   
1   3     2     34                 0.0                      0.0   
2   4     2     23                 0.0                      0.0   
3   8     2     37                 0.0                    195.0   
4  10     2     39                 0.0                      0.0   

   imp_op_var39_comer_ult3  imp_op_var40_comer_ult1  imp_op_var40_comer_ult3  \
0                      0.0                      0.0                      0.0   
1                      0.0                      0.0                      0.0   
2                      0.0                      0.0                      0.0   
3                    195.0                      0.0                      0.0   
4                      0.0                      0.0                      0.0   

   imp_op_var40_efect_ult1  imp_op_var40_efect_ult3  imp_op_var40_ult1  \
0                      0.0                      0.0                0.0   
1                      0.0                      0.0                0.0   
2                      0.0                      0.0                0.0   
3                      0.0                      0.0                0.0   
4                      0.0                      0.0                0.0   

   imp_op_var41_comer_ult1  imp_op_var41_comer_ult3  imp_op_var41_efect_ult1  \
0                      0.0                      0.0                      0.0   
1                      0.0                      0.0                      0.0   
2                      0.0                      0.0                      0.0   
3                    195.0                    195.0                      0.0   
4                      0.0                      0.0                      0.0   

   imp_op_var41_efect_ult3  imp_op_var41_ult1  imp_op_var39_efect_ult1  \
0                      0.0                0.0                      0.0   
1                      0.0                0.0                      0.0   
2                      0.0                0.0                      0.0   
3                      0.0              195.0                      0.0   
4                      0.0                0.0                      0.0   

   imp_op_var39_efect_ult3  imp_op_var39_ult1  imp_sal_var16_ult1  ind_var1_0  \
0                      0.0                0.0                 0.0           0   
1                      0.0                0.0                 0.0           0   
2                      0.0                0.0                 0.0           0   
3                      0.0              195.0                 0.0           0   
4                      0.0                0.0                 0.0           0   

   ind_var1  ind_var2_0  ind_var2  ind_var5_0  ind_var5  ind_var6_0  ind_var6  \
0         0           0         0           1         0           0         0   
1         0           0         0           1         0           0         0   
2         0           0         0           1         1           0         0   
3         0           0         0           1         1           0         0   
4         0           0         0           1         0           0         0   

   ind_var8_0  ind_var8  ind_var12_0  ind_var12  ind_var13_0  \
0           0         0            0          0            0   
1           0         0            0          0            1   
2           0         0            0          0            0   
3           0         0            0          0            0   
4           0         0            1          1            0   

   ind_var13_corto_0  ind_var13_corto  ind_var13_largo_0  ind_var13_largo  \
0                  0                0                  0                0   
1                  1                1                  0                0   
2                  0                0                  0                0   
3                  0                0                  0                0   
4   

In [7]:
# Últimas linhas do dataset
df.tail()

ID  var3  var15  imp_ent_var16_ult1  imp_op_var39_comer_ult1  \
76015  151829     2     48                 0.0                      0.0   
76016  151830     2     39                 0.0                      0.0   
76017  151835     2     23                 0.0                      0.0   
76018  151836     2     25                 0.0                      0.0   
76019  151838     2     46                 0.0                      0.0   

       imp_op_var39_comer_ult3  imp_op_var40_comer_ult1  \
76015                      0.0                      0.0   
76016                      0.0                      0.0   
76017                      0.0                      0.0   
76018                      0.0                      0.0   
76019                      0.0                      0.0   

       imp_op_var40_comer_ult3  imp_op_var40_efect_ult1  \
76015                      0.0                      0.0   
76016                      0.0                      0.0   
76017                      0.0                      0.0   
76018                      0.0                      0.0   
76019                      0.0                      0.0   

       imp_op_var40_efect_ult3  imp_op_var40_ult1  imp_op_var41_comer_ult1  \
76015                      0.0                0.0                      0.0   
76016                      0.0                0.0                      0.0   
76017                      0.0                0.0                      0.0   
76018                      0.0                0.0                      0.0   
76019                      0.0                0.0                      0.0   

       imp_op_var41_comer_ult3  imp_op_var41_efect_ult1  \
76015                      0.0                      0.0   
76016                      0.0                      0.0   
76017                      0.0                      0.0   
76018                      0.0                      0.0   
76019                      0.0                      0.0   

       imp_op_var41_efect_ult3  imp_op_var41_ult1  imp_op_var39_efect_ult1  \
76015                      0.0                0.0                      0.0   
76016                      0.0                0.0                      0.0   
76017                      0.0                0.0                      0.0   
76018                      0.0                0.0                      0.0   
76019                      0.0                0.0                      0.0   

       imp_op_var39_efect_ult3  imp_op_var39_ult1  imp_sal_var16_ult1  \
76015                      0.0                0.0                 0.0   
76016                      0.0                0.0                 0.0   
76017                      0.0                0.0                 0.0   
76018                      0.0                0.0                 0.0   
76019                      0.0                0.0                 0.0   

       ind_var1_0  ind_var1  ind_var2_0  ind_var2  ind_var5_0  ind_var5  \
76015           0         0           0         0           1         0   
76016           0         0           0         0           1         0   
76017           0         0           0         0           1         1   
76018           0         0           0         0           1         1   
76019           0         0           0         0           1         0   

       ind_var6_0  ind_var6  ind_var8_0  ind_var8  ind_var12_0  ind_var12  \
76015           0         0           0         0            0          0   
76016           0         0           0         0            1          1   
76017           0         0           0         0            0          0   
76018           0         0           0         0            0          0   
76019           0         0           0         0            0          0   

       ind_var13_0  ind_var13_corto_0  ind_var13_corto  ind_var13_largo_0  \
76015            0                  0                0                  0   
76016            0                  0             

In [8]:
# Verificando o shape dos dados
df.shape

(76020, 371)

###### São 371 variáveis presentes nesta base de dados. A última coluna é a variável target

In [9]:
# A base não possui variáveis categóricas
df.dtypes[df.dtypes == 'object']

Series([], dtype: object)

In [10]:
# A primeira coluna é o ID, ou seja, um código identificador do cliente ou do registro e 
# pode ser excluída sem prejuízo para a modelo.

del df['ID']

In [11]:
# A base de dados não possui valores nulos
df[df.isnull().values]

Empty DataFrame
Columns: [var3, var15, imp_ent_var16_ult1, imp_op_var39_comer_ult1, imp_op_var39_comer_ult3, imp_op_var40_comer_ult1, imp_op_var40_comer_ult3, imp_op_var40_efect_ult1, imp_op_var40_efect_ult3, imp_op_var40_ult1, imp_op_var41_comer_ult1, imp_op_var41_comer_ult3, imp_op_var41_efect_ult1, imp_op_var41_efect_ult3, imp_op_var41_ult1, imp_op_var39_efect_ult1, imp_op_var39_efect_ult3, imp_op_var39_ult1, imp_sal_var16_ult1, ind_var1_0, ind_var1, ind_var2_0, ind_var2, ind_var5_0, ind_var5, ind_var6_0, ind_var6, ind_var8_0, ind_var8, ind_var12_0, ind_var12, ind_var13_0, ind_var13_corto_0, ind_var13_corto, ind_var13_largo_0, ind_var13_largo, ind_var13_medio_0, ind_var13_medio, ind_var13, ind_var14_0, ind_var14, ind_var17_0, ind_var17, ind_var18_0, ind_var18, ind_var19, ind_var20_0, ind_var20, ind_var24_0, ind_var24, ind_var25_cte, ind_var26_0, ind_var26_cte, ind_var26, ind_var25_0, ind_var25, ind_var27_0, ind_var28_0, ind_var28, ind_var27, ind_var29_0, ind_var29, ind_var30_0, ind_var30, ind_var31_0, ind_var31, ind_var32_cte, ind_var32_0, ind_var32, ind_var33_0, ind_var33, ind_var34_0, ind_var34, ind_var37_cte, ind_var37_0, ind_var37, ind_var39_0, ind_var40_0, ind_var40, ind_var41_0, ind_var41, ind_var39, ind_var44_0, ind_var44, ind_var46_0, ind_var46, num_var1_0, num_var1, num_var4, num_var5_0, num_var5, num_var6_0, num_var6, num_var8_0, num_var8, num_var12_0, num_var12, num_var13_0, num_var13_corto_0, num_var13_corto, ...]
Index: []

In [12]:
# A base de dados possui registros duplicados.
# Eles serão excluídos por meio do metódo abaixo:

df = df.drop_duplicates()

# Mostra o shape dos dados após a exclusão dos registros
print(df.shape)

(71213, 370)


### Separando a variável target

In [13]:
# Variável TARGET
y = df.TARGET

In [14]:
# Variáveis de Entrada
X = df.drop('TARGET', axis = 1)

### Análise da correlação entre as variáveis

In [15]:
# corr = X.corr()
# corr.style.background_gradient(cmap='coolwarm').set_precision(2)

In [16]:
# Excluindo colunas com correlação acima de 0.70
excluir_colunas = ['num_op_var39_ult1','num_op_var41_ult3','num_op_var39_hace2','num_op_var39_ult1',
                  'num_op_var39_ult3','num_op_var39_comer_ult1','num_op_var39_comer_ult3',
                  'num_op_var41_comer_ult1','num_op_var41_comer_ult3','imp_op_var39_comer_ult3',
                  'imp_op_var41_comer_ult3','num_op_var41_ult1','imp_op_var40_comer_ult1','imp_op_var39_comer_ult1','imp_op_var40_efect_ult1','imp_op_var41_efect_ult3','imp_op_var41_ult1',
                  'imp_op_var39_ult1','num_op_var40_ult1','saldo_var40','num_op_var40_comer_ult1','num_op_var40_efect_ult1',
                  'num_op_var40_efect_ult3','num_op_var40_comer_ult3','ind_var6','ind_var8_0','ind_var8','ind_var12_0','ind_var12','ind_var13_0',
                  'ind_var13_corto_0','ind_var13_corto','imp_op_var41_efect_ult1','imp_op_var39_efect_ult1','imp_op_var39_efect_ult3','ind_var1_0',
                  'ind_var1','ind_var5','ind_var6_0','ind_var13_largo_0','ind_var13_largo','ind_var13_medio_0',
                  'ind_var13_medio','ind_var13','ind_var14_0','ind_var14','ind_var17_0','ind_var17','ind_var18_0','ind_var18','ind_var20_0','ind_var20','ind_var24_0','ind_var24',
                  'ind_var25_cte','ind_var26_0','ind_var26_cte','ind_var26','ind_var25_0','ind_var25','ind_var29_0',
                  'ind_var29','ind_var30','ind_var31_0','ind_var31','ind_var32_cte','ind_var32','ind_var33_0',
                  'ind_var33','ind_var34_0','ind_var34','ind_var37_cte','ind_var37_0','ind_var37','ind_var39_0',
                  'ind_var40_0','ind_var40','ind_var41_0','ind_var5_0','ind_var32_0','ind_var44_0','ind_var44','num_var1_0','num_var1','num_var4',
                   'num_var5_0','num_var5','num_var6_0','num_var6','num_var8_0','num_var8','num_var12_0',
                  'num_var12','num_var13_0','num_var13_corto_0','num_var13_corto','num_var13_largo_0',
                  'num_var13_largo','num_var13_medio_0','num_var13_medio','num_var13','num_var17_0',
                  'num_var17','num_var18_0','num_var18','num_var24_0','num_var24','num_var26_0','num_var26',
                  'num_var25_0','num_var25','num_var29_0','num_var29','num_var30_0','num_var30','num_var32_0',
                  'num_var32','num_var33_0','num_var33','num_var34_0','num_var34','num_var37_med_ult2','num_var37_0','num_var40','saldo_var1',
                  'saldo_var6','saldo_var12','saldo_var13_corto','saldo_var13_largo','saldo_var13_medio',
                  'saldo_var13','saldo_var17','saldo_var18','saldo_var24','saldo_var29','saldo_var30',
                  'saldo_var31','saldo_var33','saldo_var34','saldo_var42','saldo_var44']
X.drop(excluir_colunas, axis = 1, inplace = True) 

In [17]:
# Excluindo colunas com correlação acima de 0.70
excluir_colunas = ['imp_aport_var17_hace3','imp_aport_var33_hace3','imp_aport_var33_ult1','imp_var7_emit_ult1',
                  'imp_compra_var44_hace3','imp_compra_var44_ult1','imp_reemb_var33_ult1','imp_trasp_var17_in_hace3',
                  'imp_trasp_var17_out_ult1','imp_trasp_var33_in_hace3','imp_trasp_var33_in_ult1',
                  'imp_trasp_var33_out_ult1','imp_venta_var44_hace3','imp_venta_var44_ult1','ind_var7_emit_ult1',
                  'ind_var7_recib_ult1','ind_var10_ult1','ind_var10cte_ult1','ind_var9_cte_ult1','ind_var9_ult1',
                  'num_aport_var17_ult1','num_aport_var33_ult1','num_var7_emit_ult1','num_var7_recib_ult1',
                  'num_compra_var44_ult1','num_var22_hace2','num_var22_hace3','num_med_var22_ult3',
                  'num_med_var45_ult3','num_meses_var5_ult3','num_meses_var13_corto_ult3','num_meses_var13_largo_ult3',
                  'num_meses_var13_medio_ult3','num_meses_var39_vig_ult3','num_meses_var44_ult3','num_op_var41_efect_ult1',
                  'num_op_var41_efect_ult3','num_op_var39_efect_ult1','ind_var39','num_op_var40_hace2','num_op_var41_hace3','num_op_var39_hace3','num_var40_0',
                  'num_var41_0','num_var39','saldo_var25','saldo_var5','saldo_var8','delta_imp_aport_var13_1y3',
                  'delta_imp_aport_var17_1y3','delta_imp_aport_var33_1y3','delta_imp_compra_var44_1y3',
                  'delta_imp_reemb_var13_1y3','delta_imp_reemb_var17_1y3','delta_imp_reemb_var33_1y3',
                  'delta_imp_trasp_var17_in_1y3','delta_imp_trasp_var17_out_1y3','delta_imp_trasp_var33_in_1y3',
                  'delta_imp_trasp_var33_out_1y3','delta_imp_venta_var44_1y3','delta_num_aport_var13_1y3',
                  'delta_num_aport_var17_1y3','delta_num_aport_var33_1y3','delta_num_compra_var44_1y3',
                  'delta_num_reemb_var13_1y3','delta_num_reemb_var17_1y3','delta_num_reemb_var33_1y3',
                  'delta_num_trasp_var17_in_1y3','delta_num_trasp_var17_out_1y3','delta_num_trasp_var33_in_1y3',
                   'delta_num_trasp_var33_out_1y3','delta_num_venta_var44_1y3','imp_amort_var18_ult1',
                   'imp_reemb_var13_ult1','imp_reemb_var17_hace3','imp_reemb_var17_ult1','num_var44','delta_imp_amort_var34_1y3','imp_aport_var13_hace3','imp_trans_var37_ult1',
                   'num_aport_var13_hace3','num_aport_var33_hace3','num_var22_ult3','num_meses_var33_ult3',
                   'num_var43_emit_ult1','num_var43_recib_ult1','num_trasp_var33_in_ult1','num_var45_ult1',
                   'num_var45_ult3','saldo_medio_var5_hace2','saldo_medio_var5_ult1','saldo_medio_var8_ult3',
                   'saldo_medio_var12_hace2','saldo_medio_var12_ult1','saldo_medio_var13_corto_hace2',
                   'saldo_medio_var13_corto_ult1','saldo_medio_var13_largo_hace2','saldo_medio_var13_largo_ult1',
                   'saldo_medio_var13_medio_hace2','saldo_medio_var13_medio_ult1','saldo_medio_var17_hace2',
                   'saldo_medio_var17_ult1','saldo_medio_var29_ult1','saldo_medio_var33_hace2','saldo_medio_var33_ult1',
                   'saldo_medio_var44_hace2','saldo_medio_var44_ult1','saldo_medio_var17_ult3','num_meses_var29_ult3',
                   'num_var31','num_var20','imp_op_var40_ult1','imp_op_var40_ult1','imp_op_var40_comer_ult3']
X.drop(excluir_colunas, axis = 1, inplace = True) 

In [18]:
# Shape dos dados após a exclusão
X.shape

(71213, 116)

In [19]:
# Correlação entre as variáveis atualizada
# corr = X.corr()
# corr.style.background_gradient(cmap='coolwarm').set_precision(2)

### Divisão em dados de treino / teste

In [20]:
# Importando a biblioteca necessária
from sklearn.model_selection import train_test_split

In [21]:
# Dividindo os dados em treino e teste
X_treino, X_teste, y_treino, y_teste = train_test_split(X, 
                                                        y, 
                                                        test_size = 0.3, 
                                                        random_state = 1000,
                                                        stratify = df.TARGET)

In [22]:
# Resultado da divisão dos dados
print(X_treino.shape, X_teste.shape, y_treino.shape, y_teste.shape)

(49849, 116) (21364, 116) (49849,) (21364,)


### Balanceamento de Classes

In [23]:
# Variável Target está desbalanceada, ou seja,
# Existem mais exemplos de clientes satisfeitos (0)
# a clientes não satisfeitos com o banco (1)

y_treino.value_counts()

0    47879
1     1970
Name: TARGET, dtype: int64

In [24]:
# Para balancear as classes, será utilizado o método over_sampling
# que cria dados sintéticos para equilibrar as quantidades de 0 e 1

import imblearn
from imblearn.over_sampling import SMOTE

In [25]:
over_sampler = SMOTE(k_neighbors = 2)

In [26]:
X_btreino, y_btreino = over_sampler.fit_resample(X_treino, y_treino)

In [27]:
# Classes balanceadas
y_btreino.value_counts()

0    47879
1    47879
Name: TARGET, dtype: int64

### Algoritmo KNN

In [28]:
# Lista de possíveis valores de K
vizinhos = list(range(1, 20, 2))

In [29]:
# Lista para os scores
cv_scores = []

In [30]:
from sklearn.neighbors import KNeighborsClassifier
from sklearn.model_selection import cross_val_score

# Validação cruzada para determinar o melhor valor de k
for k in vizinhos:
    knn = KNeighborsClassifier(n_neighbors = k)
    scores = cross_val_score(knn, X_btreino, y_btreino, cv = 5, scoring = 'accuracy')
    cv_scores.append(scores.mean()) 

In [31]:
# Ajustando o erro de classificação
erro = [1 - x for x in cv_scores]

In [32]:
# Determinando o melhor valor de k (com menor erro)
optimal_k = vizinhos[erro.index(min(erro))]
print('O valor ideal de k é %d' % optimal_k)

O valor ideal de k é 1


In [33]:
# Criando o modelo
modelo_knn = KNeighborsClassifier(n_neighbors = optimal_k)

In [34]:
# Treinando o modelo
modelo_knn.fit(X_btreino, y_btreino)

KNeighborsClassifier(n_neighbors=1)

In [35]:
import joblib

# Salva o modelo em disco
with open('modelo_knn_v2.pkl', 'wb') as pickle_file:
      joblib.dump(modelo_knn, 'modelo_knn_v2.pkl') 

In [36]:
# Previsões
y_pred_knn = modelo_knn.predict(X_teste)

In [37]:
from sklearn.metrics import roc_curve, auc, roc_auc_score, confusion_matrix
from sklearn.metrics import accuracy_score

# Matriz de Confusão
confusion_matrix(y_teste, y_pred_knn)

array([[18473,  2046],
       [  665,   180]])

In [38]:
# Previsão de probabilidade da classe positiva
y_pred_proba_knn = modelo_knn.predict_proba(X_teste)[:,1]

In [39]:
# Curva ROC nos dados e previsões em teste
roc_auc_knn = roc_auc_score(y_teste, y_pred_knn)
print(roc_auc_knn)

0.5566526449291765


In [40]:
# Curva ROC nos dados e previsões em teste
fpr_knn, tpr_knn, thresholds = roc_curve(y_teste, y_pred_proba_knn)

In [41]:
# AUC em teste
auc_knn = auc(fpr_knn, tpr_knn)
print(auc_knn)

0.5566526449291765


In [42]:
# Acurácia em teste
acuracia_knn = accuracy_score(y_teste, y_pred_knn)
print(acuracia_knn)

0.8731042875865943


### Importa arquivo de teste (Kaggle)

In [43]:
# Importando os dados de teste (fornecido pelo Kaggle)
df_teste = pd.read_csv("test.csv", sep = ",", encoding_errors='ignore')

In [44]:
# Variável ID
pred_final = pd.DataFrame(df_teste.ID)

In [45]:
# Selecionando as mesmas colunas utilizadas para treinamento do modelo
y_teste = df_teste[X.columns]

In [46]:
# Obtendo o shape dos dados
y_teste.shape

(75818, 116)

In [47]:
# Previsões em teste
y_pred_final = modelo_knn.predict(y_teste)

In [48]:
y_pred_final

array([0, 0, 0, ..., 0, 0, 0])

In [49]:
# Preparando dados finais
pred_final['TARGET'] = y_pred_final

In [50]:
pred_final

ID  TARGET
0           2       0
1           5       0
2           6       0
3           7       0
4           9       0
...       ...     ...
75813  151831       0
75814  151832       0
75815  151833       0
75816  151834       0
75817  151837       0

[75818 rows x 2 columns]

In [51]:
# Gerando arquivo csv
pred_final.to_csv("submission.csv", index = False)